compiler/rustc_data_structures/src/vec_cache.rs

```diff
@@ -0,0 +1,266 @@
+//! VecCache maintains a mapping from K -> (V, I) pairing. K and I must be roughly u32-sized, and V
+//! must be Copy.
+//!
+//! VecCache supports efficient concurrent put/get across the key space, with write-once semantics
+//! (i.e., a given key can only be put once). Subsequent puts will panic.
+//!
+//! This is currently used for query caching.
+
+use std::fmt::Debug;
+use std::marker::PhantomData;
+use std::mem::offset_of;
+use std::sync::atomic::{AtomicPtr, AtomicU32, Ordering};
+
+use rustc_index::Idx;
+
+use crate::sync::AtomicUsize;
+
+struct Slot<V> {
+    // We never construct &Slot<V> so it's fine for this to not be in an UnsafeCell.
+    value: V,
+    // This is both an index and a once-lock.
+    //
+    // 0: not yet initialized.
+    // 1: lock held, initializing.
+    // 2..u32::MAX - 2: initialized.
+    index_and_lock: AtomicU32,
+}
+
+impl<K: Idx, V, I> Default for VecCache<K, V, I> {
+    fn default() -> Self {
+        VecCache {
+            buckets: Default::default(),
+            key: PhantomData,
+            len: Default::default(),
+            present: Default::default(),
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+struct SlotIndex {
+    bucket_idx: usize,
+    entries: usize,
+    index_in_bucket: usize,
+}
+
+impl SlotIndex {
+    #[inline]
+    fn from_index(idx: u32) -> Self {
+        let mut bucket = idx.checked_ilog2().unwrap_or(0) as usize;
+        let entries;
+        let running_sum;
+        if bucket <= 11 {
+            entries = 1 << 12;
+            running_sum = 0;
+            bucket = 0;
+        } else {
+            entries = 1 << bucket;
+            running_sum = entries;
+            bucket = bucket - 11;
+        }
+        SlotIndex { bucket_idx: bucket, entries, index_in_bucket: idx as usize - running_sum }
+    }
+
+    #[inline]
+    unsafe fn get<V: Copy>(&self, buckets: &[AtomicPtr<Slot<V>>; 21]) -> Option<(V, u32)> {
+        // SAFETY: `bucket_idx` is ilog2(u32).saturating_sub(11), which is at most 21, i.e.,
+        // in-bounds of buckets. See `from_index` for computation.
+        let bucket = unsafe { buckets.get_unchecked(self.bucket_idx) };
+        let ptr = bucket.load(Ordering::Acquire);
+        // Bucket is not yet initialized: then we obviously won't find this entry in that bucket.
+        if ptr.is_null() {
+            return None;
+        }
+        // SAFETY: Follows from preconditions on `buckets` and `self`.
+        let slot = unsafe { ptr.add(self.index_in_bucket) };
+
+        // SAFETY:
+        let index_and_lock =
+            unsafe { &*slot.byte_add(offset_of!(Slot<V>, index_and_lock)).cast::<AtomicU32>() };
```

Comment:
Why do we need to `byte_add` a struct's offset here? I think this can be done with `addr_of!`.

Reply:
Doesn't something like `addr_of!((*slot).index_and_lock)` require that all of `slot` is readable at the appropriate types, i.e., that `value` is initialized here?

Reply:
Hm, poking at more docs, I think I misremembered. But I actually think we don't need even addr_of here, we have a fully initialized field (index_and_lock) and so we can just directly write this expression (`&(*slot).index_and_lock`).

Reply:
You didn't misremember, we loosened the rule in part because having to do this dance with `byte_add` is very silly.

Reply:
And we loosened in general? So I'm correct that the new form (without addr of entirely) is fine?

The dance with byte_add is/was somewhat easier for me to reason about, mostly just around reasoning where the load actually occurs (both literally and in terms of available promises for the compiler to rely on)... but anyway, no need to litigate that here, just wish we had some clearer way to check this (docs or runtime clarity...)

Reply:
The documentation of `addr_of!` says this:

> The expr in addr_of!(expr) is evaluated as a place expression, but never loads from the place or requires the place to be dereferenceable. This means that addr_of!((*ptr).field) still requires the projection to field to be in-bounds, using the same rules as [offset](https://doc.rust-lang.org/stable/std/primitive.pointer.html#method.offset). However, addr_of!(*ptr) is defined behavior even if ptr is null, dangling, or misaligned.

https://doc.rust-lang.org/stable/std/ptr/macro.addr_of.html#safety

There's no load, the place expression is just an inbounds address computation.